src/Service/Bulk/BulkUserCreationService.php

```diff
@@ -0,0 +1,327 @@
+            $this->assignmentIngester->ingest($assignmentDtoCollection);
+        }
+
+        return $noOfUsersCreated;
+    }
+
+    private function createUserPassword(): string
+    {
+        return substr(str_shuffle('0123456789abcdefghijklmnopqrstvwxyz'), 0, 8);
+    }
+
+    private function createUserDto(string $username, string $userPassword, string $userGroupId): UserDto
+    {
+        return new UserDto(
+            $username,
+            $this->passwordHasher->hashPassword(new User(), $userPassword),
+            $userGroupId ? $userGroupId : null
+        );
+    }
+
+    private function createAssignmentDto(
+        int $lineKey,
+        string $username
+    ): AssignmentDto {
+
+        return new AssignmentDto(Assignment::STATE_READY, $lineKey, $username);
+    }
+
+    private function getFindLineItemCriteria(): FindLineItemCriteria
+    {
+        $criteria = new FindLineItemCriteria();
+
+        if (!empty($this->lineItemIds)) {
+            $criteria->addLineItemIds(...$this->lineItemIds);
+        }
+
+        if (!empty($this->lineItemSlugs)) {
+            $criteria->addLineItemSlugs(...$this->lineItemSlugs);
+        }
+
+        return $criteria;
+    }
+
+    /**
+     * Function to get all slug data
+     * @throws LineItemNotFoundException
+     */
+    private function getAllLineItemSlugs(): void
+    {
+        $lineItems = $this->lineItemRepository->findAllAsCollection()->jsonSerialize();
+        if (empty($lineItems)) {
+            throw new LineItemNotFoundException('No line items were found in database.');
+        }
+        $this->generateSlugData($lineItems);
+    }
+
+    private function getLastUserAssignedToLineItems(
+        array $lineItemSlugs
+    ): array {
+        $userNameIncArr = [];
+        foreach ($lineItemSlugs as $slugKey => $slug) {
+            $assignment = $this->assignmentRepository->findByLineItemId($slugKey);
+            $userNameIncArr[$slug] = self::DEFAULT_USERNAME_INCREMENT_VALUE;
+            if (!empty($assignment)) {
+                $userInfo = $assignment->getUser()->getUsername();
+                $userNameArr = explode('_', (string)$userInfo);
+                $userNameLastNo = preg_match('/^\d+$/', end($userNameArr))
+                    ? (int)end($userNameArr)
+                    : self::DEFAULT_USERNAME_INCREMENT_VALUE;
+                $userNameIncArr[$slug] = $userNameLastNo;
+            }
+        }
+
+        return $userNameIncArr;
+    }
+
+    private function getLoadBalanceGroupID(string $groupPrefix): array
+    {
+        $totalInstance = $this->ltiInstanceRepository->findAllAsCollection()->count();
+        $targetId = 1;
+        $groupIds = [];
+        while ($targetId <= $totalInstance) {
+            $random = substr(md5(random_bytes(10)), 0, 10);
+            $groupIds[] = sprintf($groupPrefix . '_%s', $random);
+            $targetId++;
+        }
+        return $groupIds;
+    }
```

Comment:
This cam be moved to a separated component and would help you to reduce this class complexity